[Do not merge] Fixes the undefined method empty? for nil class error at the time of Windows ssh connection.

Vasu1105 opened the pull request:
Signed-off-by: Vasu1105 <vasundhara.jagdale@msystechnologies.com>

<!--- Provide a short summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail, what problems does it solve? -->
Windows ssh with `pty: true` results in Error `undefined method empty? for nil class` 
Overriding `pty` value and set it to false before `request_pty` call on ssh channel resolves the issue.
## Related Issue
<!--- If you are suggesting a new feature or change, please create an issue first -->
<!--- Please link to the issue, discourse, or stackoverflow here: -->
#473
## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New content (non-breaking change)
- [ ] Breaking change (a content change which would break existing functionality or processes)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I have read the **CONTRIBUTING** document.

lamont-granquist commented on lib/train/platforms/detect/helpers/os_windows.rb:
```diff
@@ -2,7 +2,12 @@ module Train::Platforms::Detect::Helpers
+      # setting pty to false to execute platform dection code as setting pty to true results in getting unwanted charters in response.
+      # resetting the same once output is gathered.
+
+      set_pty_false if @backend.class.to_s == "Train::Transports::SSH::Connection"
```
That is some really brittle tight coupling.

Likely run_command needs to take some kind of `may_need_pty_false: true` flag (note: but some better name) which  `Train::Transports::SSH::Connection` implements and which the other connection types ignore.

marcparadise commented on lib/train/platforms/detect/helpers/os_windows.rb:
```diff
@@ -2,7 +2,12 @@ module Train::Platforms::Detect::Helpers
+      # resetting the same once output is gathered.
+
+      set_pty_false if @backend.class.to_s == "Train::Transports::SSH::Connection"
       res = @backend.run_command("cmd.exe /c ver")
```
What was the output we were getting here when PTY was enabled?

Vasu1105 replied:
```
"\e[?7l\e[H\e[2J\e[?25l\e[3;1H\e[37m\e[40mBootDevice=\\Device\\HarddiskVolume1\e[4;1HBuildNumber=9600\e[5;1HBuildType=Multiprocessor Free\e[6;1HCaption=Microsoft Windows Server 2012 R2 Datacenter\e[7;1HCodeSet=1252\e[8;1HCountryCode=1\e[9;1HCreationClassName=Win32_OperatingSystem\e[10;1HCSCreationClassName=Win32_ComputerSystem\e[11;1HCSDVersion=\e[12;1HCSName=INS\e[13;1HCurrentTimeZone=0\e[14;1HDataExecutionPrevention_32BitApplications=TRUE\e[15;1HDataExecutionPrevention_Available=TRUE\e[16;1HDataExecutionPrevention_Drivers=TRUE\e[17;1HDataExecutionPrevention_SupportPolicy=3\e[18;1HDebug=FALSE\e[19;1HDescription=\e[20;1HDistributed=FALSE\e[21;1HEncryptionLevel=256\e[22;1HForegroundApplicationBoost=2\e[23;1HFreePhysicalMemory=2563936\e[24;1HFreeSpaceInPagingFiles=1019932\e[24;1H\n\e[JFreeVirtualMemory=3489724\e[24;1H\n\e[JInstallDate=20190624070359.000000+000\e[24;1H\n\e[JLargeSystemCache=\e[24;1H\n\e[JLastBootUpTime=20190624070147.414150+000\e[24;1H\n\e[JLocalDateTime=20190702045319.668000+000\e[24;1H\n\e[JLocale=0409\e[24;1H\n\e[JManufacturer=Microsoft Corporation\e[24;1H\n\e[JMaxNumberOfProcesses=4294967295\e[24;1H\n\e[JMaxProcessMemorySize=137438953344\e[24;1H\n\e[JMUILanguages={\"en-US\"}\e[24;1H\n\e[JName=Microsoft Windows Server 2012 R2 Datacenter|C:\\Windows|\\Device\\Harddisk0\\Pa\e[24;1H\n\e[Jrtition1\e[24;1H\n\e[JNumberOfLicensedUsers=0\e[24;1H\n\e[JNumberOfProcesses=45\e[24;1H\n\e[JNumberOfUsers=3\e[24;1H\n\e[JOperatingSystemSKU=8\e[24;1H\n\e[JOrganization=\e[24;1H\n\e[JOSArchitecture=64-bit\e[24;1H\n\e[JOSLanguage=1033\e[24;1H\n\e[JOSProductSuite=400\e[24;1H\n\e[JOSType=18\e[24;1H\n\e[JOtherTypeDescription=\e[24;1H\n\e[JPAEEnabled=\e[24;1H\n\e[JPlusProductID=\e[24;1H\n\e[JPlusVersionNumber=\e[24;1H\n\e[JPortableOperatingSystem=FALSE\e[24;1H\n\e[JPrimary=TRUE\e[24;1H\n\e[JProductType=3\e[24;1H\n\e[JRegisteredUser=Windows User\e[24;1H\n\e[JSerialNumber=00253-50000-00000-AA442\e[24;1H\n\e[JServicePackMajorVersion=0\e[24;1H\n\e[JServicePackMinorVersion=0\e[24;1H\n\e[JSizeStoredInPagingFiles=1048576\e[24;1H\n\e[JStatus=OK\e[24;1H\n\e[JSuiteMask=400\e[24;1H\n\e[JSystemDevice=\\Device\\HarddiskVolume1\e[24;1H\n\e[JSystemDirectory=C:\\Windows\\system32\e[24;1H\n\e[JSystemDrive=C:\e[24;1H\n\e[JTotalSwapSpaceSize=\e[24;1H\n\e[JTotalVirtualMemorySize=4980328\e[24;1H\n\e[JTotalVisibleMemorySize=3931752\e[24;1H\n\e[JVersion=6.3.9600\e[24;1H\n\e[JWindowsDirectory=C:\\Windows\e[24;1H\n\e[J\n\e[J\n\e[J\n\e[J\e[?12l\e[?25h\e[?7h"
```

marcparadise commented:
THe thing we want to express is "for windows detection, do not enable PTY".   Instead of worrying about class type matching inside of OS detection, could we implement this as an optional argument to `run_command`? Something like `BaseConnection#run_command(cmd, override_opts = nil, &block)`

`BaseConnection`'s `run_command` would pass `override_opts` into its call to `run_command_via_connection`.  This would require updating the connection handlers to universally accept override_opts (default {}). 

The changes to platform detection here would then look like: 
`@backend.run_command("cmd.exe /c ver", pty: false)`
 
and

` @backend.run_command("wmic os get * /format:list")`

lamont-granquist commented on lib/train/plugins/base_connection.rb:
```diff
@@ -118,10 +118,9 @@ def platform
-      return run_command_via_connection(cmd, &data_handler) unless cache_enabled?(:command)
-
-      @cache[:command][cmd] ||= run_command_via_connection(cmd, &data_handler)
+    def run_command(cmd, override_opts = {}, &data_handler)
```
these should be `**`'d and probably should just be `**opts` instead of `override_opts = {}`.

btm commented:
@Vasu1105 you may need to rebase or run `chefstyle -a` to fix some new style changes that we recently released to fix the Travis CI failure.

Vasu1105 commented:
@btm I have fixed all chefstyle in another PR here https://github.com/inspec/train/pull/474 so that I will keep this PR specific. I am looking into @tas50 and @lamont-granquist comments on that PR. Once that is done we can merge that first and then I will rebase this branch.

zenspider commented on lib/train/transports/ssh_connection.rb:
```diff
@@ -273,18 +273,20 @@ def to_s
 
-        if @transport_options[:pty]
+        # In case of Windows SSH we are checking opts[:pty] flag
+        if opts[:pty].nil? ? opts[:pty] : @transport_options[:pty]
```
This logic looks backwards to me. If `opts[:pty]` is nil, use it, otherwise use `transpart_options`... I think it should be:

```ruby
if opts[:pty] || @transport_options[:pty]
```

Unless we're trying to use trinary logic (which I'll want to shoot down as well, but separately)? Am I misunderstanding something?

Vasu1105 replied:
Thank you @zenspider. You are right in a hurry I have accidentally removed !. 
Second we have to using ternary logic as we want to overwrite default @transport_options[:pty] flag and use that in the if condition ahead. Hope this answers your queries.

clintoncwolfe commented:
@Vasu1105 can you rebase this to resolve the conflicted file above? Additionally, are you able to address the remaining feedback?  If not, we'll need to close this in a few days...

Vasu1105 commented:
@clintoncwolfe I have addressed the feedbacks. But we have hold this as we were fixing the issue from chef side itself and now as the the winrm is separated out in train-winrm. We will address it there if it still exist. So I am closing this now.
